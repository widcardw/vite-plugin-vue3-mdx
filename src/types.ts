import type { FilterPattern } from 'vite'
import type { PluggableList } from 'unified'
import type { Options as RemarkRehypeOptions } from 'remark-rehype'

interface Options {
  /**
   * Whether to keep jsx.
   *
   * @default true
   */
  jsx?: boolean

  /**
   * Format of the files to be processed.
   *
   * @default 'mdx'
   */
  format?: 'md' | 'mdx' | 'detect'

  /**
   * Whether to compile to a whole program or a function body.
   *
   * @default 'program'
   */
  outputFormat?: 'program' | 'function-body'

  /**
   * Extensions (with `.`) for markdown.
   */
  mdExtentions?: string[]

  /**
   * Extensions (with `.`) for MDX.
   */
  mdxExtensions?: string[]

  /**
   * List of recma (esast, JavaScript) plugins.
   */
  recmaPlugins?: PluggableList

  /**
   * List of remark (mdast, markdown) plugins.
   */
  remarkPlugins?: PluggableList

  /**
   * List of rehype (hast, HTML) plugins.
   */
  rehypePlugins?: PluggableList

  /**
   * Options to pass through to `remark-rehype`.
   */
  remarkRehypeOptions?: RemarkRehypeOptions

  /**
   * Whether to compile to dynamic import expressions.
   *
   * @default false
   */
  useDynamicImport?: boolean

  /**
   * Resolve imports (and `export â€¦ from`, and `import.meta.url`) from this URL
   */
  baseUrl?: string

  /**
   * Whether to add extra information to error messages in generated code
   *
   * @default false
   */
  development?: boolean

  /**
   * Class names for wrapper div
   *
   * @warning vueJsx cannot parse `<MDXLayout><ChildComponents /></MDXLayout>`
   * into `<div class="markdown-body"><ChildCompoents /></div>`. Instead, the
   * `MDXLayout` element should contain a `slot`, which may be suitable for vue.
   *
   * @default false
   */
  wrapperClasses?: false | string | string[]

  SourceMapGenerator?: any

  /**
   * Custom tranformations apply before and after the markdown transformation
   */
  transforms?: {
    before?: (code: string, id: string) => string
    after?: (code: string, id: string) => string
  }

  /**
   * In order to solve the problem of vue router
   *
   * @default true
   */
  addDisplayName?: boolean

  include?: FilterPattern
  exclude?: FilterPattern
}

interface ResolvedOptions extends Required<Options> {
  wrapperClasses: string | false
}

export {
  Options,
  ResolvedOptions,
}
