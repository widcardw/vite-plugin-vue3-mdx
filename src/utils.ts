function joinClasses(c: false | string | string[]) {
  if (typeof c === 'string')
    return c
  else if (Array.isArray(c))
    return c.filter((i?: string) => i).join(' ')
  return false
}

function addDisplayName(raw: string, id: string) {
  return raw.replace('export default MDXContent', `
MDXContent.displayName = '${id}';
export default MDXContent`.trim())
}

function addLayoutWrapperClasses(raw: string, classes: string) {
  return `
export default function Layout({__mdx_children__}) {
  return <div class="${classes}">{__mdx_children__}</div>;
}

${raw}`
}

export {
  joinClasses,
  addDisplayName,
  addLayoutWrapperClasses,
}
