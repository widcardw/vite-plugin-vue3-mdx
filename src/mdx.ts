import type { TransformResult } from 'vite'
import { compile } from '@mdx-js/mdx'
import type { ResolvedOptions } from './types'
import { addDisplayName, addLayoutWrapperClasses } from './utils'

function createMDXCompiler(options: ResolvedOptions) {
  return async (id: string, raw: string): Promise<TransformResult> => {
    raw = raw.trimStart()

    if (options.transforms.before)
      raw = options.transforms.before(raw, id)

    let code = String(await compile(raw, options))

    if (options.wrapperClasses)
      code = addLayoutWrapperClasses(code, options.wrapperClasses)

    return {
      code: options.addDisplayName ? addDisplayName(code, id) : code,
      map: { mappings: '' } as any,
    }
  }
}

export {
  createMDXCompiler,
}
