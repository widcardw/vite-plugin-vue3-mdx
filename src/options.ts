import type { Options, ResolvedOptions } from './types'
import { joinClasses } from './utils'

function resolveOptions(userOptions: Options): ResolvedOptions {
  const defaultOptions: ResolvedOptions = {
    jsx: true,
    format: 'detect',
    outputFormat: 'program',
    mdExtentions: ['.md', '.markdown', '.mdown', '.mkdn', '.mkd', '.mdwn', '.mkdown', '.ron'],
    mdxExtensions: ['.mdx'],
    recmaPlugins: [],
    remarkPlugins: [],
    rehypePlugins: [],
    remarkRehypeOptions: {},
    wrapperClasses: 'markdown-body',
    transforms: {},
    baseUrl: '.',
    development: false,
    SourceMapGenerator: undefined,
    useDynamicImport: false,
    include: null,
    exclude: null,
  }

  const options = { ...defaultOptions, ...userOptions }
  options.wrapperClasses = joinClasses(options.wrapperClasses)

  return options as ResolvedOptions
}

export {
  resolveOptions,
}
