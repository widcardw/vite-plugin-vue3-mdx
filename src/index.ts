import type { Plugin } from 'vite'
import { createFilter } from 'vite'
import { createMDXCompiler } from './mdx'
import { resolveOptions } from './options'
import type { Options } from './types'

function VitePluginVue3Mdx(userOptions: Options = {}): Plugin {
  const options = resolveOptions(userOptions)

  const mdxToJsx = createMDXCompiler(options)

  const filter = createFilter(
    (userOptions.mdExtentions || []).concat(userOptions.mdxExtensions || []) || /\.mdx?$/,
    userOptions.exclude,
  )
  return {
    name: 'vite-plugin-vue3-mdx',
    transform(raw, id) {
      if (!filter(id))
        return

      try {
        return mdxToJsx(id, raw)
      }
      catch (e: any) {
        this.error(e)
      }
    },
    async handleHotUpdate(ctx) {
      if (!filter(ctx.file))
        return

      const defaultRead = ctx.read
      ctx.read = async function () {
        return mdxToJsx(ctx.file, await defaultRead()).code
      }
    },
  }
}

export default VitePluginVue3Mdx
