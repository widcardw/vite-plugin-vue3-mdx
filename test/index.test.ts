import { describe, expect, it } from 'vitest'
import { createMDXCompiler } from '../src/mdx'
import { resolveOptions } from '../src/options'
import { saveToExported } from './utils'

describe('should', () => {
  const mdxToJsx = createMDXCompiler(resolveOptions({ wrapperClasses: ['ma', 'max-w-60rem'] }))

  it('exported', async () => {
    const res = await mdxToJsx('hi', '#hi')
    expect(res.code).toMatchSnapshot()

    const res2 = await mdxToJsx('uni', `
import { ref } from 'vue'

export const count = ref(0)

export const Button = () => (
  <button onClick={() => count.value++}>Click me</button>
)

hi {count.value} <Button />
    `)

    expect(res2.code).toMatchSnapshot()

    await saveToExported(res2.code, 'MdxFile')
  })
})
