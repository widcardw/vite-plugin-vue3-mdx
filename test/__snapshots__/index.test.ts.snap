// Vitest Snapshot v1

exports[`should > exported 1`] = `
"/*@jsxRuntime automatic @jsxImportSource react*/
function _createMdxContent(props) {
  const _components = Object.assign({
    p: \\"p\\"
  }, props.components);
  return <_components.p>{\\"#hi\\"}</_components.p>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
MDXContent.displayName = '0123';
export default MDXContent;
"
`;

exports[`should > exported 2`] = `
"/*@jsxRuntime automatic @jsxImportSource react*/
import {ref} from \\"vue\\";
export const count = ref(0);
export const Button = () => <button onClick={() => count.value++}>Click me</button>;
function _createMdxContent(props) {
  const _components = Object.assign({
    p: \\"p\\"
  }, props.components);
  return <_components.p>{\\"hi \\"}{count.value}{\\" \\"}<Button /></_components.p>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
MDXContent.displayName = 'uni';
export default MDXContent;
"
`;
